pkg/vsphere/toolbox/hgfs/protocol.go

```diff
@@ -0,0 +1,681 @@
+		err := binary.Write(buf, binary.LittleEndian, &r.Capabilities[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return buf.Bytes(), nil
+}
+
+// UnmarshalBinary implements the encoding.BinaryUnmarshaler interface
+func (r *RequestCreateSessionV4) UnmarshalBinary(data []byte) error {
+	buf := bytes.NewBuffer(data)
+
+	fields := []*uint32{
+		&r.NumCapabilities,
+		&r.MaxPacketSize,
+		&r.Flags,
+		&r.Reserved,
+	}
+
+	for _, p := range fields {
+		err := binary.Read(buf, binary.LittleEndian, p)
+		if err != nil {
+			return err
+		}
+	}
+
+	for i := uint32(0); i < r.NumCapabilities; i++ {
+		var cap Capability
+		err := binary.Read(buf, binary.LittleEndian, &cap)
+		if err != nil {
+			return err
+		}
+
+		r.Capabilities = append(r.Capabilities, cap)
+	}
+
+	return nil
+}
+
+// ReplyCreateSessionV4 as defined in hgfsProto.h:HgfsReplyCreateSessionV4
+type ReplyCreateSessionV4 struct {
+	SessionID       uint64
+	NumCapabilities uint32
+	MaxPacketSize   uint32
+	IdentityOffset  uint32
+	Flags           uint32
+	Reserved        uint32
+	Capabilities    []Capability
+}
+
+// MarshalBinary implements the encoding.BinaryMarshaler interface
+func (r *ReplyCreateSessionV4) MarshalBinary() ([]byte, error) {
+	buf := new(bytes.Buffer)
+
+	fields := []interface{}{
+		&r.SessionID,
+		&r.NumCapabilities,
+		&r.MaxPacketSize,
+		&r.IdentityOffset,
+		&r.Flags,
+		&r.Reserved,
+	}
+
+	for _, p := range fields {
+		err := binary.Write(buf, binary.LittleEndian, p)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for i := uint32(0); i < r.NumCapabilities; i++ {
+		err := binary.Write(buf, binary.LittleEndian, &r.Capabilities[i])
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return buf.Bytes(), nil
+}
+
+// UnmarshalBinary implements the encoding.BinaryUnmarshaler interface
+func (r *ReplyCreateSessionV4) UnmarshalBinary(data []byte) error {
+	buf := bytes.NewBuffer(data)
+
+	fields := []interface{}{
+		&r.SessionID,
+		&r.NumCapabilities,
+		&r.MaxPacketSize,
+		&r.IdentityOffset,
+		&r.Flags,
+		&r.Reserved,
+	}
+
+	for _, p := range fields {
+		err := binary.Read(buf, binary.LittleEndian, p)
+		if err != nil {
+			return err
+		}
+	}
+
+	for i := uint32(0); i < r.NumCapabilities; i++ {
+		var cap Capability
+		err := binary.Read(buf, binary.LittleEndian, &cap)
+		if err != nil {
+			return err
+		}
+
+		r.Capabilities = append(r.Capabilities, cap)
+	}
+
+	return nil
+}
+
+// RequestDestroySessionV4 as defined in hgfsProto.h:HgfsRequestDestroySessionV4
+type RequestDestroySessionV4 struct {
+	Reserved uint64
+}
+
+// ReplyDestroySessionV4 as defined in hgfsProto.h:HgfsReplyDestroySessionV4
+type ReplyDestroySessionV4 struct {
+	Reserved uint64
+}
+
+// FileName as defined in hgfsProto.h:HgfsFileName
+type FileName struct {
+	Length uint32
+	Name   string
+}
+
+// MarshalBinary implements the encoding.BinaryMarshaler interface
+func (f *FileName) MarshalBinary() ([]byte, error) {
+	name := f.Name
+	f.Length = uint32(len(f.Name))
```

Comment:
Assuming Go's string is ok there, as os.File.Name() is also a string type (https://golang.org/pkg/os/#FileInfo).